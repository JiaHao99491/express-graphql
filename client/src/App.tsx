import "bootstrap/dist/css/bootstrap.css";
import { AddProduct } from "./AddProduct";
import { useQuery } from '@apollo/client';
import { CATEGORIES_PRODUCT } from "./query";
import { ProductList } from "./ProductList";
import { useState } from "react";
import { Product } from "./types";
import { ProductDetail } from "./ProductDetail";

const App = () => {
    const [product, setProduct] = useState<Product>()
    const {loading, error, data} = useQuery(CATEGORIES_PRODUCT)
    if(error){
        return <p>加载发生错误</p>
    }
    if(loading){
        return <p>加载中</p>
    }
    const {getCategories, getProducts} = data
  return (
    <div className="container">
      <div className="row">
        <div className="col-md-6">
          <div className="panel panel-default">
            <div className="panel-heading">
                <AddProduct categories={getCategories}/>
            </div>
            <div className="panel-body">
                <ProductList products={getProducts}  setProduct={setProduct}/>
            </div>
          </div>
        </div>
        <div className="col-md-6">
          <div className="panel panel-default">
            <div className="panel-body">
                <ProductDetail product={product}/>
            </div>
          </div>
        </div>
      </div>
    </div>
  );
};

export default App;
