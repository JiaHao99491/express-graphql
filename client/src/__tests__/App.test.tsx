import { render, screen } from '@testing-library/react';
import '@testing-library/jest-dom';
import userEvent from '@testing-library/user-event';
import App from '../App';

jest.mock('@apollo/client', () => {
  const actual = jest.requireActual('@apollo/client');
  return {
    ...actual,
    useQuery: jest.fn(() => ({
      loading: '',
      error: '',
      data: {
        getCategories: [
          {
            __typename: 'Category',
            id: '67b83f9a33972669986ca3d4',
            name: '图书',
            products: [],
          },
        ],
        getProducts: [],
      },
    })),
    useMutation: jest.fn(() => []),
  };
});

test('demo', () => {
  expect(true).toBe(true);
});

test('渲染主页', () => {
  render(<App />);
  const productList = screen.getByText(/产品列表/);
  const productName = screen.getByText(/产品名称/);
  expect(productList).toBeInTheDocument();
  expect(productName).toBeInTheDocument();
});

test('测试用户选择产品分类', async () => {
  render(<App />);
  const productCategory = screen.getByText(/产品分类/);
  expect(productCategory).toBeInTheDocument();

  await userEvent.selectOptions(screen.getByRole('combobox'), '');
  expect(
    (screen.getByRole('option', { name: '选择分类' }) as HTMLOptionElement)
      .selected
  ).toBe(true);

  await userEvent.selectOptions(
    screen.getByRole('combobox'),
    '67b83f9a33972669986ca3d4'
  );
  expect(
    (screen.getByRole('option', { name: '图书' }) as HTMLOptionElement).selected
  ).toBe(true);
});

test('测试用户输入产品名称', async () => {
  const user = userEvent.setup();
  render(<App />);
  const productName = screen.getByRole('textbox');
  expect(productName).toBeInTheDocument();

  await user.type(productName, '红楼梦');
  expect((productName as HTMLInputElement).value).toBe('红楼梦');
});

test('快照测试', () => {
  const { asFragment } = render(<App />);
  expect(asFragment()).toMatchSnapshot();
});
