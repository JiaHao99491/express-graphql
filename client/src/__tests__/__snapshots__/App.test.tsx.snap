// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`快照测试 1`] = `
<DocumentFragment>
  <div
    class="container"
  >
    <div
      class="row"
    >
      <div
        class="col-md-6"
      >
        <div
          class="panel panel-default"
        >
          <div
            class="panel-heading"
          >
            <form>
              <div
                class="form-group"
              >
                <label>
                  产品名称
                </label>
                <input
                  class="form-control"
                  type="text"
                  value=""
                />
              </div>
              <div
                class="form-group"
              >
                <label>
                  产品分类
                </label>
                <select
                  class="form-control"
                >
                  <option
                    value=""
                  >
                    选择分类
                  </option>
                  <option
                    value="67b83f9a33972669986ca3d4"
                  >
                    图书
                  </option>
                </select>
              </div>
              <div
                class="form-group"
              >
                <input
                  class="btn btn-primary"
                  type="submit"
                />
              </div>
            </form>
          </div>
          <div
            class="panel-body"
          >
            <table
              class="table table-striped"
            >
              <caption
                class="text-center"
              >
                产品列表
              </caption>
              <thead>
                <tr>
                  <td>
                    名称
                  </td>
                  <td>
                    分类
                  </td>
                </tr>
              </thead>
              <tbody />
            </table>
          </div>
        </div>
      </div>
      <div
        class="col-md-6"
      >
        <div
          class="panel panel-default"
        >
          <div
            class="panel-body"
          />
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
